### ISSUE TITLE
APM:Copter 3.3 Extended Tuning screen value refresh does not work properly

### ISSUE THREAD

**Issue by walmis** [diydrones/apm_planner#772]
When I hit refresh, values do not update properly. Usually fields display zeroes. The only way I can get the fields to update is to go to Full Parameter List, hit refresh there and quickly go back to extended tuning, as the values are downloading.

**Comment by walmis** [comment]
This seems like a flow control issue, link gets flooded with PARAM_REQUEST requests and most of those requests get lost. I have tried proxying over UDP with MAVProxy, works better - about 80% params get returned, but when connecting over USB about 10% params get returned. Also tried connecting over UART to APM:Copter, situation is almost the same as over USB.

I'm working on a patch to queue parameter requests, and issue new requests only when requested parameters are returned.

**Comment by walmis** [comment]
Patch #1 
UAS: Queue requested parameters

```
diff --git a/src/uas/UAS.cc b/src/uas/UAS.cc
index 4d1830f..b578f4d 100644
--- a/src/uas/UAS.cc
+++ b/src/uas/UAS.cc
@@ -2638,6 +2638,10 @@ void UAS::processParamValueMsg(mavlink_message_t& msg, const QString& paramName,
     default:
         QLOG_ERROR() << "INVALID DATA TYPE USED AS PARAMETER VALUE: " << rawValue.param_type;
     } //switch (value.param_type)
+
+    if(!paramRequestQueue.isEmpty()) {
+        requestNextParamFromQueue();
+    }
 }
 
 /**
@@ -2657,11 +2661,13 @@ void UAS::requestParameter(int component, int id)
     //QLOG_DEBUG() << __FILE__ << __LINE__ << "REQUESTING PARAM RETRANSMISSION FROM COMPONENT" << component << "FOR PARAM ID" << id;
 }
 
-/**
-* Request a parameter, use parameter name to request it.
-*/
-void UAS::requestParameter(int component, const QString& parameter)
-{
+void UAS::requestNextParamFromQueue() {
+    if(paramRequestQueue.isEmpty()) return;
+
+    QPair<int,QString> pr = paramRequestQueue.front();
+    int component = pr.first;
+    QString parameter = pr.second;
+
     // Request parameter, use parameter name to request it
     mavlink_message_t msg;
     mavlink_param_request_read_t read;
@@ -2681,6 +2687,24 @@ void UAS::requestParameter(int component, const QString& parameter)
     mavlink_msg_param_request_read_encode(systemId, componentId, &msg, &read);
     sendMessage(msg);
     QLOG_DEBUG() << __FILE__ << __LINE__ << "REQUESTING PARAM RETRANSMISSION FROM COMPONENT" << component << "FOR PARAM NAME" << parameter;
+
+    paramRequestQueue.pop_front();
+}
+
+/**
+* Request a parameter, use parameter name to request it.
+*/
+void UAS::requestParameter(int component, const QString& parameter)
+{
+    QLOG_DEBUG() << "Queuing param " << parameter;
+    QPair<int,QString> p = QPair<int,QString>(component, parameter);
+
+    paramRequestQueue.append(p);
+
+    if(paramRequestQueue.size() == 1) {
+       //requestNextParamFromQueue();
+       QTimer::singleShot(0, this, SLOT(requestNextParamFromQueue()));
+    }
 }
 
 /**
diff --git a/src/uas/UAS.h b/src/uas/UAS.h
index 0d16b9d..ace7372 100644
--- a/src/uas/UAS.h
+++ b/src/uas/UAS.h
@@ -1075,8 +1075,13 @@ protected:
     quint64 lastSendTimeGPS;     ///< Last HIL GPS message sent
     quint64 lastSendTimeSensors;
 
+    QList< QPair<int, QString> >  paramRequestQueue;
+
+
 
 protected slots:
+    void requestNextParamFromQueue();
+
     /** @brief Write settings to disk */
     void writeSettings();
     /** @brief Read settings from disk */
```

**Comment by walmis** [comment]
Patch #2:
CopterPidConfig: Disable fields if parameters are not fetched properly. Add Z Accel D field.

```
diff --git a/src/ui/configuration/CopterPidConfig.cc b/src/ui/configuration/CopterPidConfig.cc
index d6a4d04..8c5abe6 100644
--- a/src/ui/configuration/CopterPidConfig.cc
+++ b/src/ui/configuration/CopterPidConfig.cc
@@ -82,6 +82,7 @@ CopterPidConfig::CopterPidConfig(QWidget *parent) : AP2ConfigWidget(parent)
 
     m_nameToBoxMap["ACCEL_Z_P"] = ui.accelZPSpinBox;
     m_nameToBoxMap["ACCEL_Z_I"] = ui.accelZISpinBox;
+    m_nameToBoxMap["ACCEL_Z_D"] = ui.accelZDSpinBox;
     m_nameToBoxMap["ACCEL_Z_IMAX"] = ui.accelZIMAXSpinBox;
     m_nameToBoxMap["ACCEL_Z_FILT_HZ"] = ui.accelZFiltHzSpinBox;
 
@@ -273,17 +274,21 @@ void CopterPidConfig::parameterChanged(int uas, int component, QString parameter
     if (m_nameToBoxMap.contains(parameterName))
     {
         m_nameToBoxMap[parameterName]->setValue(value.toDouble());
+        m_nameToBoxMap[parameterName]->setEnabled(true);
     }
     else if (parameterName == "TUNE_HIGH")
     {
         ui.ch6MaxSpinBox->setValue(value.toDouble() / 1000.0);
+        ui.ch6MaxSpinBox->setEnabled(true);
     }
     else if (parameterName == "TUNE_LOW")
     {
         ui.ch6MinSpinBox->setValue(value.toDouble() / 1000.0);
+        ui.ch6MinSpinBox->setEnabled(true);
     }
     else if (parameterName == "TUNE")
     {
+        ui.ch6OptComboBox->setEnabled(true);
         for (int i=0;i<m_ch6ValueToTextList.size();i++)
         {
             if (m_ch6ValueToTextList[i].first == value.toInt())
@@ -294,6 +299,7 @@ void CopterPidConfig::parameterChanged(int uas, int component, QString parameter
     }
     else if (parameterName == "CH7_OPT")
     {
+        ui.ch7OptComboBox->setEnabled(true);
         for (int i=0;i<m_ch78ValueToTextList.size();i++)
         {
             if (m_ch78ValueToTextList[i].first == value.toInt())
@@ -304,6 +310,7 @@ void CopterPidConfig::parameterChanged(int uas, int component, QString parameter
     }
     else if (parameterName == "CH8_OPT")
     {
+        ui.ch8OptComboBox->setEnabled(true);
         for (int i=0;i<m_ch78ValueToTextList.size();i++)
         {
             if (m_ch78ValueToTextList[i].first == value.toInt())
diff --git a/src/ui/configuration/CopterPidConfig.ui b/src/ui/configuration/CopterPidConfig.ui
index 1624a5d..c2a0b6e 100644
--- a/src/ui/configuration/CopterPidConfig.ui
+++ b/src/ui/configuration/CopterPidConfig.ui
@@ -7,7 +7,7 @@
     <x>0</x>
     <y>0</y>
     <width>677</width>
-    <height>575</height>
+    <height>588</height>
    </rect>
   </property>
   <property name="sizePolicy">
@@ -113,6 +113,9 @@
                <layout class="QVBoxLayout" name="verticalLayout_30">
                 <item>
                  <widget class="QDoubleSpinBox" name="posZPSpinBox">
+                  <property name="enabled">
+                   <bool>false</bool>
+                  </property>
                   <property name="decimals">
                    <number>4</number>
                   </property>
@@ -190,6 +193,9 @@
                <layout class="QVBoxLayout" name="verticalLayout_27">
                 <item>
                  <widget class="QDoubleSpinBox" name="velZPSpinBox">
+                  <property name="enabled">
+                   <bool>false</bool>
+                  </property>
                   <property name="decimals">
                    <number>4</number>
                   </property>
@@ -237,6 +243,9 @@
        </item>
        <item row="4" column="0">
         <widget class="QPushButton" name="writePushButton">
+         <property name="enabled">
+          <bool>true</bool>
+         </property>
          <property name="text">
           <string>Write Params</string>
          </property>
@@ -371,6 +380,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_10">
               <item>
                <widget class="QDoubleSpinBox" name="rateRollPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -387,6 +399,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateRollISpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -403,6 +418,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateRollDSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -419,6 +437,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateRollIMAXSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="minimum">
                  <double>-10000.000000000000000</double>
                 </property>
@@ -429,6 +450,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateRollFiltHzSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="suffix">
                  <string> Hz</string>
                 </property>
@@ -553,6 +577,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_24">
               <item>
                <widget class="QDoubleSpinBox" name="rateYawPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -569,6 +596,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateYawISpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -585,6 +615,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateYawDSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -601,6 +634,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateYawIMAXSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="minimum">
                  <double>-10000.000000000000000</double>
                 </property>
@@ -611,6 +647,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="rateYawFiltHzSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="suffix">
                  <string> Hz</string>
                 </property>
@@ -648,6 +687,9 @@
            </item>
            <item>
             <widget class="QComboBox" name="ch6OptComboBox">
+             <property name="enabled">
+              <bool>false</bool>
+             </property>
              <property name="sizePolicy">
               <sizepolicy hsizetype="Preferred" vsizetype="Fixed">
                <horstretch>0</horstretch>
@@ -687,6 +729,9 @@
            </item>
            <item>
             <widget class="QDoubleSpinBox" name="ch6MinSpinBox">
+             <property name="enabled">
+              <bool>false</bool>
+             </property>
              <property name="maximumSize">
               <size>
                <width>200</width>
@@ -726,6 +771,9 @@
            </item>
            <item>
             <widget class="QDoubleSpinBox" name="ch6MaxSpinBox">
+             <property name="enabled">
+              <bool>false</bool>
+             </property>
              <property name="maximumSize">
               <size>
                <width>200</width>
@@ -762,6 +810,9 @@
            </item>
            <item>
             <widget class="QComboBox" name="ch7OptComboBox">
+             <property name="enabled">
+              <bool>false</bool>
+             </property>
              <property name="sizePolicy">
               <sizepolicy hsizetype="Minimum" vsizetype="Fixed">
                <horstretch>0</horstretch>
@@ -801,6 +852,9 @@
            </item>
            <item>
             <widget class="QComboBox" name="ch8OptComboBox">
+             <property name="enabled">
+              <bool>false</bool>
+             </property>
              <property name="sizePolicy">
               <sizepolicy hsizetype="Minimum" vsizetype="Fixed">
                <horstretch>0</horstretch>
@@ -879,6 +933,13 @@
                 </property>
                </widget>
               </item>
+              <item alignment="Qt::AlignRight">
+               <widget class="QLabel" name="label_17">
+                <property name="text">
+                 <string>D</string>
+                </property>
+               </widget>
+              </item>
               <item>
                <widget class="QLabel" name="label_45">
                 <property name="text">
@@ -905,6 +966,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_33">
               <item>
                <widget class="QDoubleSpinBox" name="accelZPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -921,6 +985,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="accelZISpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -936,7 +1003,20 @@
                </widget>
               </item>
               <item>
+               <widget class="QDoubleSpinBox" name="accelZDSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
+                <property name="decimals">
+                 <number>4</number>
+                </property>
+               </widget>
+              </item>
+              <item>
                <widget class="QDoubleSpinBox" name="accelZIMAXSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -950,6 +1030,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="accelZFiltHzSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="suffix">
                  <string> Hz</string>
                 </property>
@@ -1024,6 +1107,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_4">
               <item>
                <widget class="QDoubleSpinBox" name="stabilPitchPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="sizePolicy">
                  <sizepolicy hsizetype="Fixed" vsizetype="Fixed">
                   <horstretch>0</horstretch>
@@ -1084,6 +1170,9 @@
           </item>
           <item>
            <widget class="QDoubleSpinBox" name="posXYPSpinBox">
+            <property name="enabled">
+             <bool>false</bool>
+            </property>
             <property name="decimals">
              <number>4</number>
             </property>
@@ -1148,6 +1237,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_6">
               <item>
                <widget class="QDoubleSpinBox" name="stabilYawPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1243,6 +1335,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_21">
               <item>
                <widget class="QDoubleSpinBox" name="velXYPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1259,6 +1354,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="velXYISpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1275,6 +1373,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="velXYIMAXSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1288,6 +1389,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="velXYFiltHzSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="suffix">
                  <string> Hz</string>
                 </property>
@@ -1402,6 +1506,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_12">
               <item>
                <widget class="QDoubleSpinBox" name="ratePitchPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1418,6 +1525,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="ratePitchISpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1434,6 +1544,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="ratePitchDSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>4</number>
                 </property>
@@ -1450,6 +1563,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="ratePitchIMAXSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="minimum">
                  <double>-10000.000000000000000</double>
                 </property>
@@ -1460,6 +1576,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="ratePitchFiltHzSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="suffix">
                  <string> Hz</string>
                 </property>
@@ -1549,6 +1668,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_2">
               <item>
                <widget class="QDoubleSpinBox" name="stabilRollPSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="minimumSize">
                  <size>
                   <width>0</width>
@@ -1663,6 +1785,9 @@
              <layout class="QVBoxLayout" name="verticalLayout_36">
               <item>
                <widget class="QDoubleSpinBox" name="wpNavSpeedSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>0</number>
                 </property>
@@ -1676,6 +1801,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="wpNavRadiusSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>0</number>
                 </property>
@@ -1689,6 +1817,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="wpNavSpeedUpSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>0</number>
                 </property>
@@ -1702,6 +1833,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="wpNavSpeedDownSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>0</number>
                 </property>
@@ -1715,6 +1849,9 @@
               </item>
               <item>
                <widget class="QDoubleSpinBox" name="wpNavLoiterSpeedSpinBox">
+                <property name="enabled">
+                 <bool>false</bool>
+                </property>
                 <property name="decimals">
                  <number>0</number>
                 </property>
```

**Comment by billbonney** [comment]
How are you connecting and which OS?

Also https://help.github.com/articles/fork-a-repo and create a branch* and push your changes to your branch, I can then review your changes much easier.  You can even create a pull request to get your fixes merged to the repo.

*when you make changes in your fork always work on a branch and not master, it will make your life easier.

Thanks for looking at this issue :+1:

**Comment by walmis** [comment]
Thanks, I have created a pull request.

I tried connecting with the following methods:
1. UART connection with FTDI cable, 57kbaud
2. USB ttyACM
3. Over MAVProxy UDP.
All of which have the same problem more or less. Refresh button grabs at most 5-7 parameters, all others are lost. I think the request are lost on APM side, since the MAVLink loop runs there at 50hz IIRC, therefore the MAVLink UART buffer fills up with requests faster than APM loop can handle.

OS is Ubuntu 14.10
